src/shared/pkcs11-util.c

```diff
@@ -632,96 +668,276 @@ int pkcs11_token_find_private_key(
-                                       "Failed to initialize object find call: %s", sym_p11_kit_strerror(rv));
+        if (r < 0)
+                return log_error_errno(r, "Failed to find private key");
 
-        rv = m->C_FindObjects(session, objects, ELEMENTSOF(objects), &n_objects);
-        rv2 = m->C_FindObjectsFinal(session);
-        if (rv != CKR_OK)
-                return log_error_errno(SYNTHETIC_ERRNO(EIO),
-                                       "Failed to find objects: %s", sym_p11_kit_strerror(rv));
-        if (rv2 != CKR_OK)
-                return log_error_errno(SYNTHETIC_ERRNO(EIO),
-                                       "Failed to finalize object find call: %s", sym_p11_kit_strerror(rv));
         if (n_objects == 0)
                 return log_error_errno(SYNTHETIC_ERRNO(ENOENT),
-                                       "Failed to find selected private key suitable for decryption on token.");
+                                       "Failed to find selected private key suitable for decryption or derivation on token.");
         if (n_objects > 1)
                 return log_error_errno(SYNTHETIC_ERRNO(ENOTUNIQ),
-                                       "Configured private key URI matches multiple keys, refusing.");
+                                       "Configured private key URI matches multiple derivation keys, refusing.");
 
         *ret_object = objects[0];
         return 0;
 }
 
-int pkcs11_token_decrypt_data(
+static inline void pkcs11_token_destroy_session_object(
+                CK_FUNCTION_LIST *m,
+                CK_SESSION_HANDLE session,
+                CK_OBJECT_HANDLE handle) {
+        assert(m);
+
+        CK_RV rv = m->C_DestroyObject(session, handle);
+        if (rv != CKR_OK)
+                log_debug_errno(SYNTHETIC_ERRNO(EIO), "Failed to destroy object, ignoring.");
+}
+
+#if HAVE_OPENSSL
+static int pkcs11_token_ecdh_derive_shared_secret(
+                CK_FUNCTION_LIST *m,
+                CK_SESSION_HANDLE session,
+                CK_OBJECT_HANDLE object,
+                EVP_PKEY *peer_key,
+                uint8_t *ret_shared_secret,
+                size_t *ret_shared_secret_len) {
+
+        assert(m);
+        assert(peer_key);
+        assert(ret_shared_secret_len);
+
+        _cleanup_free_ uint8_t *buffer = NULL;
+        size_t shared_secret_len = 0, public_data_len = 0;
+        static const CK_BBOOL yes = CK_TRUE, no = CK_FALSE;
+        CK_OBJECT_HANDLE shared_secret_handle = CK_INVALID_HANDLE;
+        static const CK_OBJECT_CLASS shared_secret_class = CKO_SECRET_KEY;
+        static const CK_KEY_TYPE shared_secret_type = CKK_GENERIC_SECRET;
+        CK_RV rv;
+        CK_ECDH1_DERIVE_PARAMS params = {
+                .kdf = CKD_NULL,
+                .pSharedData = NULL,
+                .ulSharedDataLen = 0,
+        };
+        CK_ATTRIBUTE shared_secret_template[] = {
+                { CKA_TOKEN, (void *)&no, sizeof(no) },
```

Comment:
hmm, we usually write this out as `(void*) &no` instead of `(void *)&no`...

also, can you align this please like a table with three columns? it's how we generall do things these days